src/Common/Common.Api/Gateways/GatewayClient.cs

```diff
@@ -188,6 +188,134 @@ public async Task<HttpResponseMessage> UpdateInstallerPrincipals(UpdateGatewayIn
+            UpdateGatewayClusterDatasourceRequest updateDatasourceRequest,
+            bool asIndividual)
+        {
+            var url = Invariant($"{GetODataUrlStart(true)}/gatewayClusters/{gatewayClusterId}/datasources/{gatewayClusterDatasourceId}");
+
+            var httpContent = SerializeObject(updateDatasourceRequest);
+            var response = await HttpClientInstance.PatchAsync(url, httpContent);
+            response.EnsureSuccessStatusCode();
+
+            return response;
+        }
+
+        public async Task<HttpResponseMessage> DeleteGatewayClusterDatasource(Guid gatewayClusterId, Guid gatewayClusterDatasourceId, bool asIndividual)
+        {
+            var url = Invariant($"{GetODataUrlStart(true)}/gatewayClusters/{gatewayClusterId}/datasources/{gatewayClusterDatasourceId}");
+
+            var response = await HttpClientInstance.DeleteAsync(url);
+            response.EnsureSuccessStatusCode();
+
+            return response;
+        }
+
+        public async Task<HttpResponseMessage> AddUsersToGatewayClusterDatasource(Guid gatewayClusterId, Guid gatewayClusterDatasourceId, UserAccessRightEntry datasourceUser, bool asIndividual)
+        {
+
+            var url = Invariant($"{GetODataUrlStart(true)}/gatewayClusters/{gatewayClusterId}/datasources/{gatewayClusterDatasourceId}/users");
+
+            var httpContent = SerializeObject(datasourceUser);
+            var response = await HttpClientInstance.PostAsync(url, httpContent);
+            response.EnsureSuccessStatusCode();
+
+            return response;
+        }
+
+        public async Task<IEnumerable<UserAccessRightEntry>> GetGatewayClusterDatasourceUsers(Guid gatewayClusterId, Guid gatewayClusterDatasourceId, bool asIndividual)
+        {
+            var url = Invariant($"{GetODataUrlStart(true)}/gatewayClusters/{gatewayClusterId}/datasources/{gatewayClusterDatasourceId}/users");
+
+            var response = await HttpClientInstance.GetAsync(url);
+            response.EnsureSuccessStatusCode();
+
+            var datasourceUsers = await DeserializeResponseContent<ODataResponseList<UserAccessRightEntry>>(response);
+
+            return datasourceUsers?.Value;
+        }
+
+        public async Task<HttpResponseMessage> RemoveGatewayClusterDatasourceUser(Guid gatewayClusterId, Guid gatewayClusterDatasourceId, string user, bool asIndividual)
+        {
+            var url = Invariant($"{GetODataUrlStart(true)}/gatewayClusters/{gatewayClusterId}/datasources/{gatewayClusterDatasourceId}/users/{user}");
+
+            var response = await HttpClientInstance.DeleteAsync(url);
+            response.EnsureSuccessStatusCode();
+
+            return response;
+        }
+
+        public async Task<IEnumerable<DatasourceErrorDetails>> UpdateGatewayClusterDatasourceCredentials(
+            Guid gatewayClusterId,
+            Guid gatewayClusterDatasourceId,
+            DatasourceCredentialDetails updateCredentialDetails,
```

Comment:
'use UpdateGatewayClusterDatasourceCredentialRequest' instead of 'DatasourceCredentialDetails' . When are we going to update this?

Reply:
I'd like to keep the contracts exactly the same.